src/serac/physics/tests/dynamic_thermal_adjoint.cpp

```diff
@@ -0,0 +1,329 @@
+
+  shapeDisp.Add(pertubation, shape_derivative_direction);
+
+  double qoi = 0.0;
+  thermal->outputState();
+  for (int i = 0; i < ts_info.num_timesteps; ++i) {
+    double dt = ts_info.totalTime / ts_info.num_timesteps;
+    thermal->advanceTimestep(dt);
+    thermal->outputState();
+    qoi += computeStepQoi(thermal->temperature(), dt);
+  }
+  return qoi;
+}
+
+std::tuple<double, FiniteElementDual, FiniteElementDual> computeThermalQoiAndInitialTemperatureAndShapeSensitivity(
+    axom::sidre::DataStore& data_store, const NonlinearSolverOptions& nonlinear_opts,
+    const TimesteppingOptions&                                                  dyn_opts,
+    const heat_transfer::IsotropicConductorWithLinearConductivityVsTemperature& mat, const TimeSteppingInfo& ts_info)
+{
+  auto thermal = create_nonlinear_heat_transfer(data_store, nonlinear_opts, dyn_opts, mat);
+
+  double qoi = 0.0;
+  thermal->outputState();
+  for (int i = 0; i < ts_info.num_timesteps; ++i) {
+    double dt = ts_info.totalTime / ts_info.num_timesteps;
+    thermal->advanceTimestep(dt);
+    thermal->outputState();
+    qoi += computeStepQoi(thermal->temperature(), dt);
+  }
+
+  FiniteElementDual initial_temperature_sensitivity(thermal->temperature().space(), "init_temp_sensitivity");
+  initial_temperature_sensitivity = 0.0;
+  FiniteElementDual shape_sensitivity(thermal->shapeDisplacement().space(), "shape_sensitivity");
+  shape_sensitivity = 0.0;
+
+  FiniteElementDual adjoint_load(thermal->temperature().space(), "adjoint_load");
+
+  for (int i = ts_info.num_timesteps; i > 0; --i) {
+    double             dt                      = ts_info.totalTime / ts_info.num_timesteps;
+    FiniteElementState temperature_end_of_step = thermal->loadCheckpointedTemperature(thermal->cycle());
+    computeStepAdjointLoad(temperature_end_of_step, adjoint_load, dt);
+    thermal->reverseAdjointTimestep({{"temperature", adjoint_load}});
+    shape_sensitivity += thermal->computeTimestepShapeSensitivity();
+  }
+
+  EXPECT_EQ(0, thermal->cycle());  // we are back to the start
+  auto initialConditionSensitivities     = thermal->computeInitialConditionSensitivity();
+  auto initialTemperatureSensitivityIter = initialConditionSensitivities.find("temperature");
+  SLIC_ASSERT_MSG(initialTemperatureSensitivityIter != initialConditionSensitivities.end(),
+                  "Could not find temperature in the computed initial condition sensitivities.");
+  initial_temperature_sensitivity += initialTemperatureSensitivityIter->second;
+
+  return std::make_tuple(qoi, initial_temperature_sensitivity, shape_sensitivity);
+}
+
+std::tuple<double, FiniteElementDual, FiniteElementDual> computeThermalConductivitySensitivity(
+    axom::sidre::DataStore& data_store, const NonlinearSolverOptions& nonlinear_opts,
+    const TimesteppingOptions& dyn_opts, const heat_transfer::ParameterizedLinearIsotropicConductor& mat,
+    const TimeSteppingInfo& ts_info)
+{
+  printf("a\n");
```

Comment:
I assume these print statements are for debugging. They should be removed before this is merged.

Reply:
fixed